### PR TITLE
port more WKB parsing to Qgs(Const)WkbPtr including bounds checking

### PR DESCRIPTION
**Pull request by jef-n**
Might also fix [#14204](http://hub.qgis.org/issues/14204)

### CODE REVIEW

File: src/core/geometry/qgswkbptr.h [hunk 1]
```diff
@@ -26,33 +37,50 @@
 class CORE_EXPORT QgsWkbPtr
 {
     mutable unsigned char *mP;
+    unsigned char *mStart;
+    unsigned char *mEnd;
+
+    void verify( int size ) const { Q_ASSERT( mP && mP + size <= mEnd ); if ( !mP || mP + size > mEnd ) throw QgsWkbException( "wkb access out of bounds" ); }
```
Comment (strk) [inline src/core/geometry/qgswkbptr.h]: Should Q_ASSERT be removed here ?
You can "catch throw" under gdb, if that's the aim of the assertion...

File: src/core/qgsclipper.cpp [hunk 2]
```diff
@@ -37,17 +37,14 @@ const double QgsClipper::MIN_Y = -16000;
...
-  QgsConstWkbPtr wkbPtr( wkb + 1 );
+  QgsWKBTypes::Type wkbType = wkbPtr.readHeader();
 
-  unsigned int wkbType, nPoints;
-
-  wkbPtr >> wkbType >> nPoints;
-
-  bool hasZValue = QgsWKBTypes::hasZ( static_cast< QgsWKBTypes::Type >( wkbType ) );
-  bool hasMValue = QgsWKBTypes::hasM( static_cast< QgsWKBTypes::Type >( wkbType ) );
+  int nPoints;
```
Comment (strk) [inline src/core/qgsclipper.cpp]: Why was "nPoints" changed from unsigned to signed int ?

**Comment by strk** [comment]
The test in #2751 now works with this PR, if you want to merge it here, that PR can be closed

**Comment by strk** [comment]
(actually, only after removing the Q_ASSERT)

File: src/core/qgsclipper.cpp (hunk 2 above)
Comment (jef-n) [inline src/core/qgsclipper.cpp]: just because the counters are mostly used signed elsewhere.

**Comment by jef-n** [comment]
c01b61f catches the exception in ``QgsGeometryFactory::geomFromWkb`` and returns ``nullptr``.

**Comment by strk** [comment]
Uhm, how can we tell something went wrong after QgsGeometry::fromWkb ? (it currently returns void)

**Comment by jef-n** [comment]
``QgsGeometry::geometry()`` returns ``nullptr``.

**Comment by dwadler** [comment]
It looks like these fixes may resolve the defect 14204 that I opened.

Can QgsGeometry::fromWkb be used from anywhere outside of the QGIS base code?  Changing the parameter from unsigned char * to QgsConstWkbPtr is going to break things.

Is QgsConstWkbPtr going to replace QgsWkbPtr?  

Since QgsGeometry is still going to hold the original wkb string, any use of QgsWkbPtr or native logic against QgsGeometry::asWkb is going to fail if the endian-ness is different. My proposed fix modified the wkb stored in QgsGeometry to be in the native endian format.

**Comment by jef-n** [comment]
QgsConstWkbPtr is used to read WKB (and also switch endianess) and QgsWkbPtr to create wkb (in host endianess).    We could add a default for size in the QgsConstWkbPtr constructor - but that would of course circumvent the bounds checking.

**Comment by strk** [comment]
The ideal situation would be for a single class (QgsWkbPtr)
with const methods for reading and non-const methods for writing.

**Comment by strk** [comment]
@jef-n what about exportToWkb ? And asWkb ? wkbSize ? Is there a special value of all the types returned by all the methods that signifies "this geometry is broken" ?

**Comment by strk** [comment]
@jef-n ok, now I got it, we're going after each other. What about you merge in my test and continue here while I move to something else ?

**Pull request merged**